### PR TITLE
Side 1195 consolidate loop

### PR DESCRIPTION
**Pull request by MaggieMaSS**
This is the link of the task: https://crosschx.atlassian.net/browse/SIDE-1195

### CODE REVIEW

File: ldk/javascript/examples/consolidate-loop/Patient.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
+class Patient {
+    constructor(firstName, lastName, dob, gender, telephone, email, visitReason, appointmentDate, appointmentTime){
+        this._firstName = firstName;
```
Comment (MaggieMaSS) [inline ldk/javascript/examples/consolidate-loop/Patient.js:3]: will discard "_"

File: ldk/javascript/examples/consolidate-loop/Patient.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        this._gender = gender;
+        this._telephone = telephone;
+        this._email = email;
+        this._visitReason = visitReason;
+        this._appointmentDate = appointmentDate;
+        this._appointmentTime = appointmentTime;
+    }
+
+    patientInfo = [
+        '_firstName',
```
Comment (MaggieMaSS) [inline ldk/javascript/examples/consolidate-loop/Patient.js:15]: will discard "_"?

File: ldk/javascript/examples/consolidate-loop/Patient.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
+class Patient {
```
Comment (cshields-oliveai) [inline ldk/javascript/examples/consolidate-loop/Patient.js:1]: I would research a prettier for js. I see some places you have included semicolons and some you have not.

File: ldk/javascript/examples/consolidate-loop/Patient.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+          }
+          patientRecord += this[k] || ''
+        })
+        return patientRecord
+    }
+
+    getSearchCred () {
+        return `${this._firstName} ${this._lastName}:${this._email}:${this._appointmentDate}:${this._appointmentTime}`
+    }
+  //TODO:
```
Comment (MaggieMaSS) [inline ldk/javascript/examples/consolidate-loop/Patient.js:56]: will delete todo

File: ldk/javascript/examples/consolidate-loop/Patient.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    //Check whether valid MM/dd/yyyy Date Format.
+    if (!regex.test(val)) {
+        this._dob = null;
+        throw new Error('date of birth must be as follow format: MM/DD/YYYY')
+    }
+    this._dob = val
+}
+
+  setGender(val) {
+    const genderList = ["Male", "Female"]
```
Comment (MaggieMaSS) [inline ldk/javascript/examples/consolidate-loop/Patient.js:75]: +"any"

File: ldk/javascript/examples/consolidate-loop/Patient.js [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    if (!genderList.includes(val)) {
+      throw new Error("gender need to be “Male” or “Female”") 
+    }
+    this._gender = val
+  }
+  setTelephone(val) {
```
Comment (cshields-oliveai) [inline ldk/javascript/examples/consolidate-loop/Patient.js:81]: Check if these properties are required, if they are, add some more validation to them like you had with above examples.

File: ldk/javascript/examples/consolidate-loop/README.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
+# Building and running Loop in Olive Helps
```
Comment (cshields-oliveai) [inline ldk/javascript/examples/consolidate-loop/README.md:1]: Should we update this readme to include wording about what the consolidate example loop is, how to run it, and what kind of examples it provides?

File: ldk/javascript/examples/consolidate-loop/listenHotkey.js [hunk 8]
```diff
@@ -0,0 +1,154 @@
+import { clipboard, whisper, keyboard, filesystem, network } from '@oliveai/ldk'
+import Patient from './Patient'
+
+const { TextInput, Telephone,Checkbox, ListPair, Email, Button} = whisper.WhisperComponentType; 
+
+      const hotkeys = {
```
Comment (MaggieMaSS) [inline ldk/javascript/examples/consolidate-loop/listenHotkey.js:6]: indentention

File: ldk/javascript/examples/consolidate-loop/listenHotkey.js [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+                  {
+                    label: 'First Name',
+                    onChange: (error,value) =>
+                    {
+                        if (value != null )
+                        {
+                            patient.setFirstName(value);
+                            
+                        }
+                        console.log(value, patient._firstName);
```
Comment (MaggieMaSS) [inline ldk/javascript/examples/consolidate-loop/listenHotkey.js:31]: delete log

File: ldk/javascript/examples/consolidate-loop/listenHotkey.js [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+                onClose: () => {
+                  console.log('closed Results');
+                },
+                components: [
+                  {
+                    label: 'First Name',
+                    onChange: (error,value) =>
+                    {
+                        if (value != null )
+                        {
```
Comment (MaggieMaSS) [inline ldk/javascript/examples/consolidate-loop/listenHotkey.js:27]: time to change to js stayle

File: ldk/javascript/examples/consolidate-loop/listenHotkey.js [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+                  {
+                    label: 'Appointment Time',
+                    onChange: (error,value) => 
+                    {
+                        patient.setAppointmentTime();
+                    },
+                    type: TextInput,
+                  },
+                  {
+                    // Close&clear the loop after clicking submit
```
Comment (MaggieMaSS) [inline ldk/javascript/examples/consolidate-loop/listenHotkey.js:103]: do we need to leave the comment?

File: ldk/javascript/examples/consolidate-loop/listenHotkey.js [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+                    {
+                      try {
+                        patient.verify()
+                      } catch (e) {
+                        console.error(e)
+                        return
+                      }
+                        // varify if having duplicated patients
+                        // If patients have same name and dob,use email to distinguish. 
+                      let cred1 = `${patient._firstName}:${patient._lastName}:${patient._dob}`
```
Comment (MaggieMaSS) [inline ldk/javascript/examples/consolidate-loop/listenHotkey.js:115]: extract the verification  as a new function

File: ldk/javascript/examples/consolidate-loop/listenHotkey.js [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+                    })
+
+                    if (found) {
+                      throw new Error(`patient already exist: ${cred1}:${cred2}`)
+                    }
+                      network
+                        .encode(patientRecord)
+                        .then((encodedValue) => {
+                          console.log('ENCODED: ', encodedValue)
+                          setTimeout(() => {
```
Comment (cshields-oliveai) [inline ldk/javascript/examples/consolidate-loop/listenHotkey.js:136]: I don't believe you need `setTimeout()`

File: ldk/javascript/examples/consolidate-loop/listenHotkey.js [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+                            filesystem.writeFile({
+                              path: "./PatientInfo.txt",
+                              data: encodedValue,
+                              writeOperation: filesystem.WriteOperation.append,
+                              writeMode: 0o744
+                            }).then(() => {
+                                console.log("filesystem.writeFile callback")
```
Comment (MaggieMaSS) [inline ldk/javascript/examples/consolidate-loop/listenHotkey.js:143]: Successfully saved patient record

File: ldk/javascript/examples/consolidate-loop/listenHotkey.js [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+                            }).catch(console.error)
+                          })
+                        })
+                      instance.close && instance.close()
```
Comment (cshields-oliveai) [inline ldk/javascript/examples/consolidate-loop/listenHotkey.js:147]: I believe you want `instance.close()`

File: ldk/javascript/examples/consolidate-loop/package.json [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+          {
+            "value": "./PatientInfo.txt"
+          }
+        ]
+      }
+    }
+  },
+  "dependencies": {
+    "@oliveai/ldk": "file:../..",
+    "file-system": "^2.2.2"
```
Comment (cshields-oliveai) [inline ldk/javascript/examples/consolidate-loop/package.json:29]: I don't think `file-system` ref is needed.

File: ldk/javascript/tests/default-with-config/.babelrc [hunk 17]
```diff
@@ -1,5 +1,6 @@
 {
     "plugins": ["@babel/plugin-syntax-dynamic-import"],
+
```
Comment (MaggieMaSS) [inline ldk/javascript/tests/default-with-config/.babelrc:3]: delete the new row

File: ldk/javascript/examples/consolidate-loop/UISearch.js [hunk 18]
```diff
@@ -0,0 +1,74 @@
+import { ui, whisper, filesystem, network } from '@oliveai/ldk'
+import Patient from './Patient'
+
+const { TextInput, Telephone,Checkbox, ListPair, Email, Button, Message, Link} = whisper.WhisperComponentType; 
+
+function SearchResultWhisper(rows){
```
Comment (cshields-oliveai) [inline ldk/javascript/examples/consolidate-loop/UISearch.js:6]: `searchResultWhisper`

File: ldk/javascript/examples/consolidate-loop/UISearch.js [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+                whisper.create({
+                    label: 'Consolidate Loop - Patient Search Result',
+                    onClose: () => {
+                      console.log('Patient Search Result');
+                    },
+                    components: [{
+                        header: row._firstName + ' ' + row._lastName,
+                        body: row.patientInfo.reduce((rst, k) => {
+                            return `
+${rst}
```
Comment (MaggieMaSS) [inline ldk/javascript/examples/consolidate-loop/UISearch.js:24]: check the format

File: ldk/javascript/examples/consolidate-loop/Patient.js [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+        }
+        if (!this._email) {
+            throw new Error('email is required')
+        }
+        if (!this._appointmentDate) {
+            throw new Error('appointmentDate is required')
+        }
+    }
+
+    seralize() {
```
Comment (epiccoleman) [inline ldk/javascript/examples/consolidate-loop/Patient.js:41]: minor typo here, should be "serialize"

File: ldk/javascript/examples/consolidate-loop/README.md [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+
+## Build a Loop
+To build a Loop, create a new project modeled after this example.
+
+Note that your Loop project will require a `package.json`. See "Producing Loop Compilations" in the main LDK readme for more information.
+
+All of your Loop code will begin with an `index.js` file. In this file, you define all of your Loop business logic, leveraging the LDK Aptitudes.
+
+## Transpile the Loop for Olive Helps
+From `examples/typescript-loop`
```
Comment (RusGraf) [inline ldk/javascript/examples/consolidate-loop/README.md:11]: probably need to update the path

File: ldk/javascript/examples/consolidate-loop/UISearch.js [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+import Patient from './Patient'
+
+const { TextInput, Telephone,Checkbox, ListPair, Email, Button, Message, Link} = whisper.WhisperComponentType; 
+
+function SearchResultWhisper(rows){
+    let result = []
+    console.log(result)
+    rows.forEach(row => {
+        result.push({
+            text: row._firstName + ' ' + row._lastName + ' ' + row._email,// clickable, email is the distinguisher if fn&ln are same
```
Comment (RusGraf) [inline ldk/javascript/examples/consolidate-loop/UISearch.js:11]: could use string interpolation here

File: ldk/javascript/examples/consolidate-loop/listenHotkey.js [hunk 23]
```diff
@@ -0,0 +1,154 @@
...
+                      }
+                        // varify if having duplicated patients
+                        // If patients have same name and dob,use email to distinguish. 
+                      let cred1 = `${patient._firstName}:${patient._lastName}:${patient._dob}` 
+                      let cred2 = patient._email
+                      let patientRecord = '|' + patient.seralize()
+
+                      const found = await new Promise((resolve, reject) => {
+
+                        filesystem.readFile("./PatientInfo.txt").then((data) => {
```
Comment (RusGraf) [inline ldk/javascript/examples/consolidate-loop/listenHotkey.js:121]: I would use await pattern here to reduce a mix of the .then/await to make it more readable.

File: ldk/javascript/examples/consolidate-loop/package.json [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+  "ldk": {
+    "permissions": {
+      "clipboard": {},
+      "keyboard": {},
+      "whisper": {},
+      "ui": {},
+      "filesystem": {
+        "pathGlobs": [
+          {
+            "value": "./PatientInfo.txt"
```
Comment (RusGraf) [inline ldk/javascript/examples/consolidate-loop/package.json:21]: I believe we dont need to specify glob permissions for the  ./ any more. As long as you have "filesystem": {} you should be good to create './PatientInfo.txt'. This change was made earlier this week.

**Review by richseviora — DISMISSED** [review]

**Review by RusGraf — APPROVED** [review]

**Review by BrettMugglinOlive — COMMENTED** [review]
The name of the loop probably shouldn't be "Consolidate Loop". Maybe we should consider renaming it in file names, documentation, folder names, and variables to something like "Scheduling Example" or something like that.

What do you think?

EDIT: After bringing it up with Robin, let's rename it to "Universal Example Loop"

File: ldk/javascript/examples/consolidate-loop/Patient.js [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+      data: encodedValue,
+      writeOperation: filesystem.WriteOperation.append,
+      writeMode: 0o744,
+    });
+
+    console.log('Successfully saved patient record');
+  }
+
+  async isAlreadyExist() {
+    const cred1 = `${this.firstName}:${this.lastName}:${this.dob}`;
```
Comment (BrettMugglinOlive) [inline ldk/javascript/examples/consolidate-loop/Patient.js:74]: If email is required, do we need to check if `cred1` is true in this function?

File: ldk/javascript/examples/consolidate-loop/Patient.js [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+  setLastName(val) {
+    this.lastName = val;
+  }
+
+  setDob(val) {
+    const regex = /^(0[1-9]|1[0-2])\/(0[1-9]|1\d|2\d|3[01])\/(19|20)\d{2}$/;
+
+    if (!regex.test(val)) {
+      this.dob = null;
+      console.error(new Error('date of birth must be as follow format: MM/DD/YYYY'));
```
Comment (BrettMugglinOlive) [inline ldk/javascript/examples/consolidate-loop/Patient.js:129]: In general, for any logs we want to use sentence case, unless the first character of the sentence is referring to a variable name. like `firstName`

File: ldk/javascript/examples/consolidate-loop/README.md [hunk 27]
```diff
@@ -0,0 +1,35 @@
+# Building and running Loop in Olive Helps
+
+### What is consolidate loop?
+
+Consolidate loop demonstrates the capabilities of Olive Helps used in conjunction with each other.
```
Comment (BrettMugglinOlive) [inline ldk/javascript/examples/consolidate-loop/README.md:5]: "The Universal Example Loop"

File: ldk/javascript/examples/consolidate-loop/README.md [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+
+### When to use consolidate loop?
```
Comment (BrettMugglinOlive) [inline ldk/javascript/examples/consolidate-loop/README.md:7]: Maybe something more like "What does it do?"

File: ldk/javascript/examples/consolidate-loop/README.md [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+
+New patient appointment scheduling as a 2-step process: 1) Patient calls, scheduling staff schedules appointment. 2) Registration staff uses this information to contact patients before their appointments to confirm the appointment and provide instructions.
```
Comment (BrettMugglinOlive) [inline ldk/javascript/examples/consolidate-loop/README.md:9]: "It's a patient appointment scheduling loop that is a 2-step process:..."

File: ldk/javascript/examples/consolidate-loop/README.md [hunk 30]
```diff
@@ -0,0 +1,35 @@
...
+
+### How to use consolidate loop?
+
+1. Hit Ctrl + N to open Consolidate Loop - Form Whisper
+2. Input patient information and submit.
+   Please note that firstname, lastname, telephone number, email and appointmentDate are required fields.
+3. Use firstname, lastname, date of birth, phone number or appointment date to search for patients using search bar, choose the right one if multiple records are returned.
```
Comment (BrettMugglinOlive) [inline ldk/javascript/examples/consolidate-loop/README.md:16]: Should either be the field names (like "firstName", "phoneNumber", etc.) or the label for the field displayed in the UI ("First Name", "Telephone", etc.)

File: ldk/javascript/examples/consolidate-loop/README.md [hunk 31]
```diff
@@ -0,0 +1,35 @@
...
+
+## Build a Loop
```
Comment (BrettMugglinOlive) [inline ldk/javascript/examples/consolidate-loop/README.md:18]: Build the Loop

File: ldk/javascript/examples/consolidate-loop/README.md [hunk 32]
```diff
@@ -0,0 +1,35 @@
...
+
+To build a Loop, create a new project modeled after this example.
```
Comment (BrettMugglinOlive) [inline ldk/javascript/examples/consolidate-loop/README.md:20]: To build the Loop

File: ldk/javascript/examples/consolidate-loop/README.md [hunk 33]
```diff
@@ -0,0 +1,35 @@
...
+
+Note that your Loop project will require a `package.json`. See "Producing Loop Compilations" in the main LDK readme for more information.
```
Comment (BrettMugglinOlive) [inline ldk/javascript/examples/consolidate-loop/README.md:22]: README

File: ldk/javascript/examples/consolidate-loop/README.md [hunk 34]
```diff
@@ -0,0 +1,35 @@
...
+
+All of your Loop code will begin with an `index.js` file. In this file, you define all of your Loop business logic, leveraging the LDK Aptitudes.
```
Comment (BrettMugglinOlive) [inline ldk/javascript/examples/consolidate-loop/README.md:24]: Can probably remove this sentence

File: ldk/javascript/examples/consolidate-loop/README.md [hunk 35]
```diff
@@ -0,0 +1,35 @@
...
+
+## Transpile the Loop for Olive Helps
```
Comment (BrettMugglinOlive) [inline ldk/javascript/examples/consolidate-loop/README.md:26]: Transpiling

File: ldk/javascript/examples/consolidate-loop/README.md [hunk 36]
```diff
@@ -0,0 +1,35 @@
...
+
+From `examples/consolidate-loop`
+
+- `npm i`
+- `npm run build`
+
+This will generate a `./dist` directory from your `index.js` that containes your transpiled Loop.
```
Comment (BrettMugglinOlive) [inline ldk/javascript/examples/consolidate-loop/README.md:33]: conatains

File: ldk/javascript/examples/consolidate-loop/searchBarListener.js [hunk 37]
```diff
@@ -0,0 +1,103 @@
...
+    label: 'Consolidate Loop - Patient Search Result',
+    onClose: () => {
+      console.log('Patient Search Result');
+    },
+    components:
+      searchResults.length > 0
+        ? searchResults
+        : [
+            {
+              body: `No results are found matching provided criteria`,
```
Comment (BrettMugglinOlive) [inline ldk/javascript/examples/consolidate-loop/searchBarListener.js:48]: No results were found

**Review by BrettMugglinOlive — APPROVED** [review]

**Pull request merged**